File: Seori/LeetCode 75/Sliding Window/1493. Longest Subarray of 1's After Deleting One Element.py
class Solution:
    def longestSubarray(self, nums: List[int]) -> int:
        answer = 0
        left, right = -1, -1
        length = len(nums)

        # [1] for문을 이용하여 window의 left, right를 세팅한다.
        for i in range(length):
            if nums[i] == 0:
                # [1-1] 0이 처음 등장했을 때
                if left == -1:
                    left = i
                # [1-2] 0이 두 번째 등장했을 때
                elif right == -1:
                    answer = i - 1
                    right = i
                    left += 1
                # [1-3] 0이 세 번 이상 등장했을 때
                else:
                    answer = max(answer, i - left - 1)
                    left = right + 1
                    right = i

        # [2] 예외처리. 0이 2개 미만인 경우와 마지막 인덱스 계산
        if right == -1:
            answer = length - 1
        elif right != length:
            answer = max(answer, length - left - 1)
        
        return answer

File: Gombang/[LeetCode75] - [Sliding Window]/1493_Longest_Subarray_of_1's_After_Deleting_One_Element.cpp
class Solution {
public:
    int longestSubarray(vector<int>& nums) {
        int left = 0;
        bool isDeleted = false;
        int result = 0;

        for (int right = 0; right < nums.size(); right++)
        {
            if (nums[right] == 0)
            {
                if (isDeleted)
                {
                    // 이미 0을 삭제한 경우에는 슬라이딩 윈도우 크기 축소
                    while (nums[left] != 0)
                        left++;

                    left++; //최종적으로 0의 다음 인덱스를 가리키고 있음.
                }
                else
                {
                    isDeleted = true;
                }
            }

            result = max(result, right - left);
        }

        return result;
    }
};